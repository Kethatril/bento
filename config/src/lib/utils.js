import objectPath from "object-path";
import { parse as shoelace } from "@kenvunz/shoelace";
import { join, resolve as res, dirname } from "path";
import bl from "browserslist";
import { existsSync, readFileSync } from "fs";
import querystring from "querystring";
import merge from "lodash/merge";

export function enhance(options) {
    return objectPath(options);
}

export function resolve(options) {
    ["input", "output"].forEach(key => {
        if (
            !options.paths ||
            !options.paths[key] ||
            options.paths[key].indexOf(options.paths.root) === 0
        )
            return;

        options.paths[key] = res(join(options.paths.root, options.paths[key]));
    });

    return options;
}

export function browsersync(options) {
    if (!options.paths || !options.paths.input) return options;
    options.browsersync = merge(
        browsersyncrc(options.paths.input),
        options.browsersync || {}
    );
    return options;
}

export function browserslist(options) {
    if (!options.paths || !options.paths.input) return options;
    options.browserslist = merge(
        browserslistrc(options.paths.input),
        options.browserslist || {}
    );
    return options;
}

export function babel(options) {
    if (!options.paths || !options.paths.input) return options;
    options.babel = merge(babelrc(options.paths.input), options.babel || {});
    return options;
}

export function transform(
    options,
    fns = [resolve, browserslist, browsersync, babel, shoelace, enhance]
) {
    for (const fn of fns) {
        options = fn(options);
    }

    return options;
}

export function select(options) {
    return function(cases) {
        return cases[options.get("env.value")];
    };
}

export function caught(e) {
    if (process.env.NODE_ENV === "test") throw e;
    console.error(e);
}

function browsersyncrc(path) {
    return rc(path, [
        "bs-config.js",
        ".browsersyncrc",
        "package.json?key=browsersync"
    ]);
}

function babelrc(path) {
    return rc(path, [".babelrc", "package.json?key=babel"]);
}

function browserslistrc(path) {
    const config = bl.findConfig(path);
    if (!config) return bl.defaults;

    if (config[process.env.NODE_ENV]) return config[process.env.NODE_ENV];
    return config.defaults;
}

function rc(path, files) {
    for (const item of files) {
        const [name, query] = item.split("?");
        const file = join(path, name);
        if (!existsSync(file)) continue;
        let options;
        if (query) options = querystring.parse(query);

        const object = JSON.parse(readFileSync(file));

        if (name !== "package.json") return object;
        return options && options.key ? object[options.key] : options;
    }

    const up = dirname(path);

    if (up !== path) {
        return rc(up, files);
    }
}
