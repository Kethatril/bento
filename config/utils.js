const { Config, environment } = require("webpack-config"),
    path = require("path"),
    objectPath = require("object-path");

let model;

module.exports.config = function(fn) {
    const config = new Config();
    return fn(config, environment);
};

module.exports.select = function(cases) {
    return cases[model.get("env.value")];
};

module.exports.read = function(key, defaultValue) {
    return model.get(key, defaultValue);
};

module.exports.set = function(opts) {
    let json = JSON.stringify(opts);
    const pre = objectPath(opts),
        matches = json.match(/"@{([\w\.]+)}["\/]/g);

    // replace any "@key" occurrences with the real value
    matches.forEach(match => {
        const key = match.replace(/[@{}\/"]/g, ""),
            value = pre.get(key);

        json = json.replace(
            new RegExp(match, "g"),
            match.substr(-1) === "/" ? `"${value}/` : JSON.stringify(value)
        );
    });

    model = objectPath(JSON.parse(json));
};

module.exports.format = function(key, value) {
    if (typeof value === "string") {
        return value;
    }
    const manifest = value;
    /**
     * Hack to prepend scripts/ or styles/ to manifest keys
     *
     * This might need to be reworked at some point.
     *
     * Before:
     *   {
     *     "main.js": "scripts/main_abcdef.js"
     *     "main.css": "styles/main_abcdef.css"
     *   }
     * After:
     *   {
     *     "scripts/main.js": "scripts/main_abcdef.js"
     *     "styles/main.css": "styles/main_abcdef.css"
     *   }
     */
    Object.keys(manifest).forEach(src => {
        const sourcePath = path.basename(path.dirname(src));
        const targetPath = path.basename(path.dirname(manifest[src]));
        if (sourcePath === targetPath) {
            return;
        }
        manifest[`${targetPath}/${src}`] = manifest[src];
        delete manifest[src];
    });
    return manifest;
};
