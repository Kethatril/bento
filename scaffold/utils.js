const pathIsInside = require("path-is-inside"),
    path = require("path");

module.exports.select = function(key, presets) {
    return function(answers) {
        return presets[answers.kind][key];
    };
};

module.exports.inside = function(name, root) {
    return function(value) {
        const resolved = path.resolve(value);
        const result = pathIsInside(resolved, root);
        return (
            result ||
            `${name} must be inside ${chalk.bold(
                root
            )}, it's currently resolved to ${chalk.bold(resolved)}`
        );
    };
};

module.exports.trailing = function(char) {
    return function(value) {
        if (value.substr(-1) !== char) value += "/";

        return value;
    };
};
