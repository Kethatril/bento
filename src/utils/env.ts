export interface EnvSelectOptions {
    [env: string]: any;
}

export function selector(env?: string): () => any {
    return function(choice: EnvSelectOptions = {}) {
        return choice[env]
            ? choice[env]
            : choice.default ? choice.default : null;
    };
}
