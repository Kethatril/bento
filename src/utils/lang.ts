export function instantiate(Cls, args?: any[]): any {
    args.unshift(Cls);
    const C = Cls.bind.apply(Cls, args);
    return new C();
}

interface SelectorOptions {
    [key: string]: any;
}

export function selector(env?: string): (choice: SelectorOptions) => any {
    return function(choice: SelectorOptions = {}) {
        return choice[env]
            ? choice[env]
            : choice.default ? choice.default : null;
    };
}
