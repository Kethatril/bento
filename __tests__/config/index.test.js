import { join } from "path";
import { scaffolder, tree, request } from "../utils";
import { build, bundle, serve } from "~/index";

jasmine.DEFAULT_TIMEOUT_INTERVAL = 120000;

describe("config/index.js", () => {
    const scaffold = scaffolder("--default spa --no-confirm");

    afterEach(() => {
        scaffolder.restore();
    });

    it("builds webpack config that matches the snapshot", () => {
        return scaffold()
            .then(dir => {
                const options = require(join(dir, "./app/config/webpack"));
                return build(options);
            })
            .then(config => {
                expect(config).toMatchSnapshot();
            });
    });

    it("bundles using webpack correctly", () => {
        return scaffold()
            .then(dir => {
                const options = require(join(dir, "./app/config/webpack"));
                options.env = Object.assign(options.env, {
                    value: "production",
                    isProduction: true
                });

                return bundle(options).then(() => dir);
            })
            .then(dir => {
                const output = tree("./public -f -l 3", dir);
                expect(output.stdout).toMatchSnapshot();
            })
            .catch(e => {
                console.error(e);
            });
    });

    it("serves using webpack-dev-server correctly", () => {
        return scaffold().then(dir => {
            const options = require(join(dir, "./app/config/webpack"));
            options.env = Object.assign(options.env, {
                value: "development",
                isProduction: false,
                isDevServer: true
            });

            options.browsersync = {
                open: false,
                logLevel: "silent"
            };

            return serve(options, (server, done) => {
                const req = request(server.app);

                Promise.all([
                    req
                        .get("/")
                        .accept("html")
                        .expect(200)
                        .test(),
                    req
                        .get("/scripts/main.js")
                        .expect("Content-Type", /javascript/)
                        .expect(200)
                        .test()
                ]).then(() => {
                    done();
                });
            });
        });
    });
});
