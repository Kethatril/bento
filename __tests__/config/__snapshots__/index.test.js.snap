// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config/index.js Server Side Application builds webpack config that matches the snapshot 1`] = `
Object {
  "context": "<cwd>/assets",
  "devtool": "cheap-module-eval-source-map",
  "entry": Object {
    "main": Array [
      "./scripts/main.js",
    ],
  },
  "module": Object {
    "rules": Array [
      Object {
        "include": Array [
          "<cwd>/assets",
        ],
        "test": /\\.js$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "presets": Array [
                Array [
                  "env",
                  Object {
                    "targets": Object {
                      "browsers": Array [
                        "> 1%",
                        "last 2 versions",
                        "Firefox ESR",
                      ],
                    },
                    "useBuiltIns": true,
                  },
                ],
              ],
            },
          },
        ],
      },
      Object {
        "include": Array [
          "<cwd>/assets",
        ],
        "test": /\\.scss$/,
        "use": Array [
          Object {
            "loader": "<cwd>/node_modules/@gladeye/bento/node_modules/extract-text-webpack-plugin/dist/loader.js",
            "options": Object {
              "omit": 1,
              "publicPath": "../",
              "remove": true,
            },
          },
          Object {
            "loader": "style-loader",
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "sourceMap": false,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "ident": "postcss",
              "plugins": [Function plugins],
              "sourceMap": false,
            },
          },
          Object {
            "loader": "resolve-url-loader",
            "options": Object {
              "sourceMap": false,
            },
          },
          Object {
            "loader": "sass-loader",
            "options": Object {
              "sourceMap": false,
            },
          },
        ],
      },
      Object {
        "include": Array [
          /node_modules/,
        ],
        "test": /\\.css$/,
        "use": Array [
          Object {
            "loader": "<cwd>/node_modules/@gladeye/bento/node_modules/extract-text-webpack-plugin/dist/loader.js",
            "options": Object {
              "omit": 1,
              "publicPath": "../",
              "remove": true,
            },
          },
          Object {
            "loader": "style-loader",
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "sourceMap": false,
            },
          },
        ],
      },
      Object {
        "include": Array [
          "<cwd>/assets",
        ],
        "test": /\\.(ttf|eot|woff2?|png|jpe?g|gif|svg)$/,
        "use": Array [
          Object {
            "loader": "file-loader",
            "options": Object {
              "name": "[path][name].[ext]",
            },
          },
        ],
      },
      Object {
        "include": /node_modules/,
        "test": /\\.(ttf|eot|woff2?|png|jpe?g|gif|svg)$/,
        "use": Array [
          Object {
            "loader": "file-loader",
            "options": Object {
              "name": "vendor/[name].[ext]",
            },
          },
        ],
      },
    ],
  },
  "output": Object {
    "filename": "scripts/[name].js",
    "path": "<cwd>/public/assets",
    "publicPath": "/assets/",
  },
  "plugins": Array [
    ExtractTextPlugin {
      "filename": "styles/[name].css",
      "id": 1,
      "options": Object {
        "disable": true,
      },
    },
    CleanWebpackPlugin {
      "options": Object {
        "allowExternal": false,
        "dry": false,
        "root": "<cwd>",
        "verbose": false,
      },
      "paths": Array [
        "<cwd>/public/assets",
      ],
    },
    ManifestPlugin {
      "opts": Object {
        "basePath": "",
        "fileName": "manifest.json",
        "filter": null,
        "generate": [Function generate],
        "map": null,
        "seed": null,
        "transformExtensions": /^(gz|map)$/i,
        "writeToFileEmit": true,
      },
    },
    DefinePlugin {
      "definitions": Object {
        "process.env": Object {
          "NODE_ENV": "\\"test\\"",
        },
      },
    },
    NamedModulesPlugin {
      "options": Object {},
    },
  ],
  "resolve": Object {
    "alias": Object {
      "@": "<cwd>/assets",
      "~": "<cwd>/assets/scripts",
    },
  },
}
`;

exports[`config/index.js Server Side Application bundles using webpack correctly 1`] = `
Object {
  "images/background.jpg": "/assets/images/background.912ae359.jpg",
  "images/nice.jpg": "/assets/images/nice.9c3c4150.jpg",
  "scripts/_manifest.js": "/assets/scripts/_manifest.d5f6378e.js",
  "scripts/_vendor.js": "/assets/scripts/_vendor.d2ca492b.js",
  "scripts/main.js": "/assets/scripts/main.9c5645b4.js",
  "styles/main.css": "/assets/styles/main.02b62d85.css",
}
`;

exports[`config/index.js Single Page Application builds webpack config that matches the snapshot 1`] = `
Object {
  "context": "<cwd>/app",
  "devtool": "cheap-module-eval-source-map",
  "entry": Object {
    "main": Array [
      "./scripts/main.js",
    ],
  },
  "module": Object {
    "rules": Array [
      Object {
        "include": Array [
          "<cwd>/app",
        ],
        "test": /\\.js$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "babelrc": false,
              "presets": Array [
                Array [
                  "env",
                  Object {
                    "targets": Object {
                      "browsers": Array [
                        "> 1%",
                        "last 2 versions",
                        "Firefox ESR",
                      ],
                    },
                    "useBuiltIns": true,
                  },
                ],
              ],
            },
          },
        ],
      },
      Object {
        "include": Array [
          "<cwd>/app",
        ],
        "test": /\\.scss$/,
        "use": Array [
          Object {
            "loader": "<cwd>/node_modules/@gladeye/bento/node_modules/extract-text-webpack-plugin/dist/loader.js",
            "options": Object {
              "omit": 1,
              "publicPath": "../",
              "remove": true,
            },
          },
          Object {
            "loader": "style-loader",
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "sourceMap": false,
            },
          },
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "ident": "postcss",
              "plugins": [Function plugins],
              "sourceMap": false,
            },
          },
          Object {
            "loader": "resolve-url-loader",
            "options": Object {
              "sourceMap": false,
            },
          },
          Object {
            "loader": "sass-loader",
            "options": Object {
              "sourceMap": false,
            },
          },
        ],
      },
      Object {
        "include": Array [
          /node_modules/,
        ],
        "test": /\\.css$/,
        "use": Array [
          Object {
            "loader": "<cwd>/node_modules/@gladeye/bento/node_modules/extract-text-webpack-plugin/dist/loader.js",
            "options": Object {
              "omit": 1,
              "publicPath": "../",
              "remove": true,
            },
          },
          Object {
            "loader": "style-loader",
          },
          Object {
            "loader": "css-loader",
            "options": Object {
              "sourceMap": false,
            },
          },
        ],
      },
      Object {
        "include": Array [
          "<cwd>/app",
        ],
        "test": /\\.(ttf|eot|woff2?|png|jpe?g|gif|svg)$/,
        "use": Array [
          Object {
            "loader": "file-loader",
            "options": Object {
              "name": "[path][name].[ext]",
            },
          },
        ],
      },
      Object {
        "include": /node_modules/,
        "test": /\\.(ttf|eot|woff2?|png|jpe?g|gif|svg)$/,
        "use": Array [
          Object {
            "loader": "file-loader",
            "options": Object {
              "name": "vendor/[name].[ext]",
            },
          },
        ],
      },
    ],
  },
  "output": Object {
    "filename": "scripts/[name].js",
    "path": "<cwd>/public",
    "publicPath": "/",
  },
  "plugins": Array [
    ExtractTextPlugin {
      "filename": "styles/[name].css",
      "id": 1,
      "options": Object {
        "disable": true,
      },
    },
    CleanWebpackPlugin {
      "options": Object {
        "allowExternal": false,
        "dry": false,
        "root": "<cwd>",
        "verbose": false,
      },
      "paths": Array [
        "<cwd>/public",
      ],
    },
    ManifestPlugin {
      "opts": Object {
        "basePath": "",
        "fileName": "manifest.json",
        "filter": null,
        "generate": [Function generate],
        "map": null,
        "seed": null,
        "transformExtensions": /^(gz|map)$/i,
        "writeToFileEmit": true,
      },
    },
    DefinePlugin {
      "definitions": Object {
        "process.env": Object {
          "NODE_ENV": "\\"test\\"",
        },
      },
    },
    NamedModulesPlugin {
      "options": Object {},
    },
    HtmlWebpackPlugin {
      "options": Object {
        "cache": true,
        "chunks": "all",
        "compile": true,
        "excludeChunks": Array [],
        "favicon": false,
        "filename": "index.html",
        "hash": false,
        "inject": true,
        "minify": false,
        "showErrors": true,
        "template": "./index.ejs",
        "title": "Webpack App",
        "xhtml": false,
      },
    },
  ],
  "resolve": Object {
    "alias": Object {
      "@": "<cwd>/app",
      "~": "<cwd>/app/scripts",
    },
  },
}
`;

exports[`config/index.js Single Page Application bundles using webpack correctly 1`] = `
Object {
  "images/background.jpg": "/images/background.912ae359.jpg",
  "images/nice.jpg": "/images/nice.9c3c4150.jpg",
  "scripts/_manifest.js": "/scripts/_manifest.2082a84b.js",
  "scripts/_vendor.js": "/scripts/_vendor.d2ca492b.js",
  "scripts/main.js": "/scripts/main.9c5645b4.js",
  "styles/main.css": "/styles/main.02b62d85.css",
}
`;
